BLE Implementation

MrSurly opened the pull request:
It's not perfect, but it's working pretty well for me.

API docs are here: https://github.com/MrSurly/micropython-esp32/blob/dev-bluetooth/esp32/bluetooth_docs.md

Merge Checklist:

## Should do:
- [X] Properties should be methods 


## Maybe:
- [ ] Move `network.Bluetooth` to top-level module `bluetooth`
- [ ] Implement the `bluepy` API as closely as possible
-- or -- 
- [ ] `service.start()` and `service.stop()` become `service.listen()` and `service.unlisten()`(?)
- [ ] Implement `select` interface for services (or should this be at the characteristic / descriptor level)

---
- [ ] Implement all functionality as separate objects (e.g. GATTC object GATTS object), while passing in  controller ID.

MrSurly committed:
esp32/Makefile: Initial changes for bluetooth
esp32/network_bluetooth.c: Initial changes for bluetooth
esp32/network_bluetooth.h:  Initial changes for bluetooth

MrSurly committed:
esp32/Makefile: Updated to include BT code from ESP-IDF
esp32/modnetwork.c: Added a Bluetooth object
esp32/network_bluetooth.*: Early development; some test code
esp32/skdconfig: Added define to enable BT

MrSurly committed:
esp32/main.c: reduced heap to 64k
esp32/network_bluetooth.c: incremental changes
esp32/sdconfig.h: Enabled bluetooth

MrSurly committed:
Merge branch 'esp32' into dev-bluetooth

MrSurly committed:
esp32/Makefile: added includes for bluetooth

MrSurly committed:
esp32/modnetwork: added declaration for bluetooth object

MrSurly committed:
esp32/network_bluetooth.h: removed this unnecessary header

MrSurly committed:
esp32/network_bluetooth.c: Added init, and params setting

Currently broken, as the call to network_bluetooth_set_ble_adv_param
in network_bluetooth_init_helper to set defaults causes a segfault

MrSurly committed:
esp32/Makefile: added a bunch more objects and headers for bluetooth

MrSurly committed:
esp32/network_bluetooth.c: Power through adding more options for ble

MrSurly committed:
esp32/network_bluetooth.c: Power through adding more options for ble

MrSurly committed:
esp32/network_bluetooth.c: Power through adding more options for ble

MrSurly committed:
esp32/Makefile: Added a whole bunch of BT objects

MrSurly committed:
esp32/network_bluetooth.c: Many, many changes, as it's early days

MrSurly committed:
esp32/network_bluetooth.c: Ensured all params checked

MrSurly committed:
esp32/network_bluetooth.c: Removed errant comment

MrSurly committed:
esp32/modesp.c: moved FAT fs to 0x200000

MrSurly committed:
esp32/modules/flashbdev.py: Increased FAT FS to 2MiB

MrSurly committed:
esp32/Makefile: Default FLASH_MODE to dio for WROOM-32

MrSurly committed:
Merge branch 'dev_issue_70' into dev-bluetooth

MrSurly committed:
esp32/main.c: Lowered heap to 16K ... =(

MrSurly committed:
esp32/sdkconfig.h: Added new needed config for BT

MrSurly committed:
esp32/network_bluetooth.c: Added ble_adv_enable function

MrSurly committed:
esp32/Makefile: added another BT object file

MrSurly committed:
esp32/network_bluetooth: Added services, characteristics, cleanup
Nowhere near ready for prime time

MrSurly committed:
esp32/network_bluetooth.c: API now more like https://goo.gl/Djt044

MrSurly committed:
esp32/network_bluetooth.c: More changes as the API evolves

MrSurly committed:
esp32/network_bluetooth.c: Got past exception in callback

MrSurly committed:
Merge branch 'esp32' into dev-bluetooth

MrSurly committed:
esp32/Makefile: many changes for new IDF; headers were clashing

MrSurly committed:
esp32/Makefile: spi_flash/include --> INC_MAIN

MrSurly committed:
esp32/network_bluetooth.c: Services working, added Bluetooth.callback()

MrSurly committed:
esp32:network_bluetooth.c: Just got characteristic callbacks working

Needs more testing

MrSurly committed:
esp32/network_bluetooth.c: Fixed bug in network_bluetooth_find_item
Thanks to @mahanet on Github

MrSurly committed:
esp32/network_bluetooth.c: Added connect/disconnect callback

MrSurly committed:
esp32/Makefile: added GATTC headers/objects

MrSurly committed:
esp32/network_bluetooth.c: Updated event debug, added gattc scanning

It just dumps logs of scan results;  I'm getting weird stuff from
the IDF somtimes.  Needs work

MrSurly committed:
esp32/network_bluetooth.c: Added callbacks for scanning

Very broken right now; getting crashes in the IDF handler.

MrSurly committed:
esp32/network_bluetooth.c: Doing callback as a ring buffer

MrSurly committed:
esp32/network_bluetooth.c: Added GIL for mallocs in IDF callback
Thanks @dpgeorge

MrSurly committed:
esp32/network_bluetooth.c: Chars() constructor now takes kw args

MrSurly committed:
esp32/network_bluetooth.c: Fixed callback handler for char write

MrSurly committed:
esp32/network_bluetooth.c: start/stop/close service is now working

MrSurly committed:
esp32/network_bluetooth.c: Added indication/notification to chars

MrSurly committed:
esp32/network_bluetooth.c: event ringbuffer --> linked list, having GC issues

MrSurly committed:
esp32/main.c: heap 16Kib -> 12kIb

MrSurly committed:
esp32/network_bluetooth.c: now using FreeRTOS queues for callback data

MrSurly committed:
esp32/bluetooth_api_proposal.txt: updated to reflect current development

MrSurly committed:
esp32/main.c: heap size back to 32Kib

MrSurly committed:
esp32/network_bluetooth.c: Many changes, much more stable now

MrSurly committed:
esp32/network_bluetooth.c: Working on adding GATTC connections

MrSurly committed:
esp32/bluetooth_api_proposal.txt: Updated api

MrSurly committed:
esp32/network_bluetooth.c: Many changes WRT GATTC

MrSurly committed:
esp32/network_bluetooth.c: GATTC nearly done; notifications still broken

MrSurly committed:
esp32/main.c: heap now 48K

MrSurly committed:
esp32/sdkconfig.h: added #define CONFIG_BT_DRAM_RELEASE 1

MrSurly committed:
esp32/network_bluetooth.c: Finished up GATTC

MrSurly committed:
esp32/bluetooth_api_proposal.txt: Removed
esp32/bluetooth_docs.md: Replaces removed proposal

MrSurly committed:
esp32/network_bluetooth.c: Added some final constants, other tweaks

MrSurly committed:
reformatted, many minor fixes

MrSurly committed:
esp32/network_bluetooth.c: Removed many unneccessary switch statements, connect now accepts bytes

MrSurly committed:
Updated docs to reflect use of bytes, added an example section

MrSurly committed:
esp32/network_bluetooth.c: Checked FIXMEs, removed whitespace, fixed adv_uuid_len

vandys commented:
Wow, this is pretty big.  If this feature should be in mainline micropython, I hope it'll land under
an ifdef.  Big bump in executable size and halving heap hurts for the non-BT apps.

MrSurly commented:
@vandys  Executable size isn't really an issue; the 4MiB flash of the WROOM 32 was [recently re-partitioned](https://github.com/micropython/micropython-esp32/pull/71) to 2MiB for code, 2MiB for Flash FS.  This weighs in at around 1480000 bytes, code-wise

I agree, heap sizing is definitely an issue, (but still better than the ESP8266!) but this arises from the IDF BT stack needing quite a bit of RAM.  I can make this a default-off config option on Monday, and we can document how to enable, with known caveats.

MrSurly committed:
esp32/Makefile: Updated to latest IDF
esp32/sdkconfig.h: Updated to latest IDF

MrSurly committed:
Merge branch 'dev-idf-1e0710f' into dev-bluetooth

MrSurly committed:
esp32/network_bluetooth.c: Updated to latest IDF
esp32/sdkconfig.h: Updated to latest IDF

MrSurly committed:
exp32/sdkconfig.h: changed defines from "y" to "1"

MrSurly committed:
esp32/Makefile: -Wno-error=array-bounds for dport_access.o
esp32/sdkconfig.h: restored CONFIG_FREERTOS_UNICORE

MrSurly committed:
Merge branch 'dev-idf-1e0710f' into dev-bluetooth-idf-dev-idf-1e0710f

MrSurly committed:
esp32/network_bluetooth.c: Fixed formatting for char print

MrSurly committed:
Merge branch 'esp32' into dev-bluetooth-idf-dev-idf-1e0710f

MrSurly committed:
esp32/modnetwork.c: Added #if around bluetooth class
esp32/mpconfigport.h Added config option for bluetooth enable

MrSurly committed:
esp32/network_bluetooth.c: disabled event debugging

MrSurly committed:
esp32/network_bluetooth.c: Added GATTS descriptors

MrSurly committed:
esp32/network_bluetooth.c: descriptors --> descrs, removed debugging prints

MrSurly committed:
esp32/network_bluetooth.c: Fixed print output for descriptors, removed "prop" from ctor

MrSurly committed:
updated docs for GATTS Descr

MrSurly commented:
Just added GATTS descriptors (@mahanet), seems to work using gatttool

MrSurly committed:
Merge branch 'esp32' into dev-bluetooth

MrSurly committed:
esp32/network_bluetooth.c: Added more parameters to scan_start

New parameters are defaulted, and shouldn't affect existing code

MrSurly committed:
esp32/bluetoot_docs.md: Updated for scan_start

bmeisels commented:
Works well all in all.
I have a problem with GATTS descriptors.
When i add one it works ok but when i add 2 descriptor to 2 different chars in a service it seems it attaches both to one of the chars and accessing them doesn't work.
example code to explain my point:
Works:
```
display_service = bluetooth.Service(UUID_DISPLAY_SERVICE)
    
    clear_screen = display_service.Char(UUID_CLEAR_SCREEN, perm=bluetooth.PERM_WRITE)
    clear_screen_description = clear_screen.Descr(ble_constants.UUID_CHAR_DESCRIPTION, value=b'clear screen',  perm=bluetooth.PERM_READ)
    clear_screen.callback(gatts_callback)

    write_line = display_service.Char(UUID_WRITE_LINE, perm=bluetooth.PERM_WRITE)
    write_line.callback(gatts_callback)

    display_service.start()
```
Doesn't work
```
display_service = bluetooth.Service(UUID_DISPLAY_SERVICE)
    
    clear_screen = display_service.Char(UUID_CLEAR_SCREEN, perm=bluetooth.PERM_WRITE)
    clear_screen_description = clear_screen.Descr(ble_constants.UUID_CHAR_DESCRIPTION, value=b'clear screen',  perm=bluetooth.PERM_READ)
    clear_screen.callback(gatts_callback)

    write_line = display_service.Char(UUID_WRITE_LINE, perm=bluetooth.PERM_WRITE)
    write_line.Descr(ble_constants.UUID_CHAR_DESCRIPTION, value=b'write line', perm=bluetooth.PERM_READ)
    write_line.callback(gatts_callback)

    display_service.start()
```

MrSurly commented:
> Works well all in all.
> I have a problem with GATTS descriptors.
> When i add one it works ok but when i add 2 descriptor to 2 different chars in a service it seems it attaches both to one of the chars and accessing them doesn't work.
example code to explain my point:

I'll check the code; in the meantime, can you attach the full code (or at least enough to make your snippets run), including all of the `UUID_*` definitions?

MrSurly commented:
Okay, this is a little weird.  Characteristics descriptors should be associated with individual characteristics, AFAIK.

But the [IDF API](http://esp-idf.readthedocs.io/en/latest/api-reference/bluetooth/esp_gatts.html#_CPPv228esp_ble_gatts_add_char_descr8uint16_tP13esp_bt_uuid_t15esp_gatt_perm_tP16esp_attr_value_tP18esp_attr_control_t) for adding a characteristic descriptor only has a _service_ handle field. I modified the code to send the _characteristic_ handle, which resulted in an error from the IDF:
```
E (7993) BT: service not created
```

I think I need to raise an issue with the IDF -- other APIs are pretty clear about this.

MrSurly committed:
esp32/network_bluetooth.c: Fix, GATTS Descr w/ multiple chars

The IDF API for adding descriptors only allows for specifying a
_service_ handle. Because of this, you must do this:

1) Add service
2) Add charactteristic
3) Add char descriptor(s)
4) Add _next_ characteristic
5) goto 3

MrSurly commented:
@bmeisels 
I just pushed a fix for this.   Note that what I'm describing below is what the C code does when you start the service -- you don't need to change anything in your Python code.

I guess you have to do this:

1) Add Service
2) Add char
3) Add char descriptors
4) Add the _next_ char
5) go to 3, until you're all out of chars.

MrSurly committed:
esp32/bluetooth_docs.md: Fixed some incorrect information WRT scan

MrSurly committed:
Merged in some parts from esp32

MrSurly committed:
Merge branch 'esp32' into dev-bluetooth

bmeisels commented:
I tested again.
this time the descriptors appear to be attached to the correct chars but they still don't always return the correct read value.

example:
```python
import network
import random

bluetooth = network.Bluetooth()

UUID_BATTERY_SERVICE = 0x180f
UUID_BATTERY_LEVEL = 0x2a19
UUID_BATTERY_LEVEL_BYTES = b'\x00\x00\x2a\x19'
UUID_BATTERY_CAPACITY = bytes(reversed(b'\xB7\xD3\xE1\x86\x64\x40\x43\xE5\xBD\x1F\x7E\x0D\x43\x25\x53\x4C'))
UUID_CHAR_DESCRIPTION = 0x2901

battery_level = None
battery_capacity = None


def gatts_callback (characteristic, event, value, userdata):
    if event == bluetooth.READ:
        print('Read')
        if characteristic is battery_level:
            print('battery_level')
            return str(random.randint(0, 100))
        if characteristic is battery_capacity:
            print('battery_capacity')
            return '2000'
        
def register_service():
    global battery_level
    global battery_capacity

    battery_service = bluetooth.Service(UUID_BATTERY_SERVICE)
    battery_level = battery_service.Char(UUID_BATTERY_LEVEL, perm=bluetooth.PERM_READ, prop=bluetooth.PROP_READ)
    battery_level.Descr(UUID_CHAR_DESCRIPTION, value=b'battery level',  perm=bluetooth.PERM_READ)
    battery_capacity = battery_service.Char(UUID_BATTERY_CAPACITY, perm=bluetooth.PERM_READ, prop=bluetooth.PROP_READ)
    battery_capacity.Descr(UUID_CHAR_DESCRIPTION, value=b'battery capacity',  perm=bluetooth.PERM_READ)
    battery_capacity.callback(gatts_callback)
    battery_level.callback(gatts_callback)
    battery_service.start()
    print('started battery service')


def bluetooth_gap_callback(bluetooth, event, data, userdata):
    if event == bluetooth.DISCONNECT:
        print("DISCONNECT")
        bluetooth.ble_adv_enable(True)


def test():
    register_service()
    bluetooth.ble_settings(adv_man_name = "esp mpy", adv_dev_name="esp mpy")
    bluetooth.ble_adv_enable(True)

    bluetooth.callback(bluetooth_gap_callback)
    
test()
```

MrSurly commented:
> but they still don't always return the correct read value.

What values are you getting?

bmeisels commented:
When i try to read the battery capacity descriptor I get 'battery level', when I try to read the battery level descriptor my app freezes(nrf connect) and after a few seconds the esp disconnects.
I have tried doing this with plain IDF c code and it works.

MrSurly committed:
esp32/network_bluetooth.c: Fixed descriptor handle enumeration

MrSurly commented:
@bmeisels:

Looks like my last fix was itself broken; wasn't dealing with handle enumeration correctly.  I've re-worked how the `service.start()` code works, and it is working for me with your example code, using `gatttool`.  

The symptoms you described are consistent with the problem I _hope_ I've fixed.

bmeisels commented:
Tested and it works.
I will continue to test more advanced configurations.
Thanks for all the help and the effort you have put into this.

MrSurly commented:
Bug reports are welcome -- I'm sure there are other issues.

dpgeorge commented:
Thanks @MrSurly for your continued work on this.  The main concern is low RAM.  We need to find a way to only use up the RAM if BT gets enabled.  For example: `esp.disable_bt()` would be used by the user to reclaim the RAM from the BT stack, and then instantiating a BT network object would rais an exception.  For this to work I think the MP GC needs to be able to handle non-contiguous regions of RAM for the heap.

dpgeorge commented:
> For example: esp.disable_bt() would be used by the user to reclaim the RAM from the BT stack,

Alternatively, a simple solution is to just have 2 builds of the firmware, one with BT enabled (and 48k heap) and one with it disabled (and with 96k heap).

MrSurly commented:
> The main concern is low RAM. We need to find a way to only use up the RAM if BT gets enabled.

Most of the RAM is eaten up by static allocations (or their effective equivalent) in the BT stack, short of getting Espressif to re-work this entirely, I don't think this is possible.

I didn't do extensive testing of how large we can make the µPy heap w/ BT -- we might be able to up it a bit. I'm not sure if there's some sort of static analysis we can do on the executable to determine how large we can make the heap.  I just did it by trial-and-error, seeing if it crashed.

> Alternatively, a simple solution is to just have 2 builds of the firmware, one with BT enabled (and 48k heap) and one with it disabled (and with 96k heap).

I started to do this by adding `#define MICROPY_BLUETOOTH (1)` to`mpconfigport.h` -- this just puts an `#if` around the code in `modnetwork.c`, but it didn't do much to reduce the executable size nor relieve the RAM issue.  I wasn't sure if the values were available in `Makefile`, but I suspect the answer is "no."

Do you have any guidance in that regard?

MrSurly committed:
esp32/bluetooth_docs.md: Added RSSI to scan results
esp32/network_bluetooth.c: Added RSSI to scan results

MrSurly committed:
esp32/network_bluetooth.c: Fixed issue with multiple chars

dpgeorge commented:
> Do you have any guidance in that regard?

Here's what we can do to configure BT support at build time:
- have a variable MICROPY_ESP32_BLUETOOTH that exists at both the makefile and C levels
- it would default to 1 (enabled) in the makefile using: `MICROPY_ESP32_BLUETOOTH ?= 1`, then it can be overridden on the command line or through a custom GNUmakefile
- the makefile would make this visible from C using:
```
ifeq ($(MICROPY_ESP32_BLUETOOTH),1)
CFLAGS += -DMICROPY_ESP32_BLUETOOTH=1
endif
```
- the makefile would wrap the addition of the BT source code (from ESP IDF) with the above ifeq/endif
- in the C code one uses `#if MICROPY_ESP32_BLUETOOTH` to configure parts that are needed for BT (and adjust the heap size in the same way)

@MrSurly does this make sense?  Do you want to attempt this?

MrSurly commented:
> @MrSurly does this make sense? Do you want to attempt this?

Yes, Yes.  My question was WRT the _correct_ way to implement =)

MrSurly committed:
esp32/Makefile: non-working attempt to match Makefile to esp32 branch

MrSurly committed:
Merge branch 'esp32' into dev-bluetooth

MrSurly committed:
esp32/Makefile: merged esp32 branch
esp32/sdkconfig.h: merged esp32 branch

MrSurly committed:
esp32/sdkconfig.h: Added new flags for BT for updated IDF

MrSurly committed:
esp32/network_bluetooth.c: Updated for IDF, fixed descr bug

MrSurly committed:
esp32/Makefile: Bluetooth now build configurable
esp32/main.c: Bluetooth now build configurable
esp32/modnetwork.c: Bluetooth now build configurable
esp32/sdkconfig.h: Bluetooth now build configurable

MrSurly commented:
@dpgeorge BT can now be enabled/disabled with `MICROPY_ESP32_BLUETOOTH=1` / `MICROPY_ESP32_BLUETOOTH=0`, defaults to enabled.  

The switch does the following:
* `Makefile`: Selectively links / compiles BT
* `main.c`: 48K heap for BT, 96K heap without
* `sdkconfig.h`: Added `#if` block for correct linker script behavior, particularly WRT `CONFIG_BT_RESERVE_DRAM`
* `modnetwork.c`: Added `#if` blocks to selectively include `Bluetooth` in `network`

Additionally, I've sync'd this PR with https://github.com/micropython/micropython-esp32

dpgeorge commented:
@MrSurly thanks for the extra commits.  I tried it out and it's not working for me, it crashes right away on the call `network.Bluetooth()` (WDT timeout, somewhere in esp_bt_controller_init).  It could be due to the latest IDF... did you try it with the 9b955f4... version of the IDF?

Update: it's crashing somewhere in btdm_controller_init

MrSurly commented:
> it crashes right away on the call ...

Well, yesterday it was working for me, and after a `make erase`, `make clean`, `make deploy`, I'm seeing what you're seeing.  Weird.  I'll dig into it.

Edit: Yes, I'm using that version if the IDF.

MrSurly committed:
esp32/Makefile: Fixed WDT crash on BT start
esp32/sdkconfig.h: Moved BT options into BT block

MrSurly commented:
> it crashes right away on the call network.Bluetooth() 

Hmm, this sounded [very familiar](https://github.com/espressif/esp-idf/issues/487), and it had to do with `CONFIG_BT_RESERVE_DRAM`, so I checked the linker script, a lo-and-behold, I wasn't passing the `#define` correctly.

In `Makefile`, this:

```
$(Q)$(CC) -I. -C -P -x c -E $(ESPCOMP)/esp32/ld/esp32.ld -o $@
```

Is now this:

```
$(Q)$(CC) -I. -C -P -x c -E $(ESPCOMP)/esp32/ld/esp32.ld -o $@ -DMICROPY_ESP32_BLUETOOTH=$(MICROPY_ESP32_BLUETOOTH)
```

It should be working, now.  Sorry to have wasted your time.  I'll be exhaustive in my testing in the future.

dpgeorge commented:
@MrSurly I tried this PR and managed to get the esp32 to scan for BLE advertisements.  But I couldn't make it act as a peripheral device.  Do you have a simple example script that just advertises, eg an Eddystone beacon?

MrSurly committed:
Minor formatting

MrSurly commented:
There's example code in the [docs](https://github.com/MrSurly/micropython-esp32/blob/dev-bluetooth/esp32/bluetooth_docs.md#heart-rate-and-gatts) -- it can act as either a HR monitor client, or a generic GATT server.

I'd have to look into Eddystone beacons to provide that specific example.  I'm certain there are use-cases that aren't covered.

dpgeorge commented:
Ok, thanks for the pointers, I now got your gatts demo working.

I think this PR should be merged because it gives a starting point to havinge BLE support on the esp32.  But there are many "conditions" to the merge of the code:
- I'm going to need to squash this entire PR (81 commits) into one commit, because the are lots of back-and-forth commits, commits with debugging stuff, etc.  The combined commits add the main file network_bluetooth.c, with changes to Makefile and sdkconfig.h to support BT.  Also very minor additions to main.c, mpnetwork.c and mpconfigport.h.  So in the end it'll be a pretty clean, self-contained commit.
- There's some stuff that's inconsistent with the rest of uPy which needs to be changed.  Eg printing the Bluetooth object shouldn't have a newline at the end, and there shouldn't be spaces between the "=" in kwargs.
- There are use of attributes (eg `char.value = 1`), and the rest of uPy-core doesn't use them, so need to think if they make sense as part of the API or should be changed to something like `char.value(1)`.
- Parts of the API could be changed to match the general network API a bit more.  Eg scanning for BLE advertisements might be better as a blocking function, with a parameter being the number of (milli)seconds to scan.
- In general the API needs further careful consideration.  In particular it's quite different from the alternative implementation in the ubluepy module in the nRF5x code: https://github.com/micropython/micropython/pull/3137 .  Both that and this implementation of BLE have their strengths/weaknesses and ultimately we want to have an API that works cleanly on esp32, nRF5x and other controllers.  It will take some time and effort to get there but what's done here is a good starting point.

@MrSurly what is your opinion of the above?

MrSurly committed:
esp32/bluetooth_docs.md: Updated docs to reflect current API

MrSurly committed:
Merge branch 'dev-bluetooth' of github.com:MrSurly/micropython-esp32 into dev-bluetooth

MrSurly commented:
> Ok, thanks for the pointers, I now got your gatts demo working.
> 
> I think this PR should be merged because it gives a starting point to havinge BLE support on the esp32. But there are many "conditions" to the merge of the code:
> 
> * I'm going to need to squash this entire PR (81 commits) into one commit, because the are lots of back-and-forth commits, commits with debugging stuff, etc. The combined commits add the main file network_bluetooth.c, with changes to Makefile and sdkconfig.h to support BT. Also very minor additions to main.c, mpnetwork.c and mpconfigport.h. So in the end it'll be a pretty clean, self-contained commit.

Sounds good.

> * There's some stuff that's inconsistent with the rest of uPy which needs to be changed. Eg printing the Bluetooth object shouldn't have a newline at the end, and there shouldn't be spaces between the "=" in kwargs.
> * There are use of attributes (eg char.value = 1), and the rest of uPy-core doesn't use them, so need to think if they make sense as part of the API or should be changed to something like char.value(1).

This is relatively easy to change.

> Parts of the API could be changed to match the general network API a bit more. Eg scanning for BLE advertisements might be better as a blocking function, with a parameter being the number of (milli)seconds to scan.

Not sure blocking would be better, especially WRT scanning, since many use cases are looking for a particular device, and scanning can take a bit of time in noisy environments -- scanning can be canceled when the device you're looking for is found.  

Would it be preferable to use threads in a blocking-scan sort of implementation?  Doesn't solve what's described above, but would allow the device to do other things while waiting for device scan.

Some features _have_ to be async (like the HR demo), since it relies on indications/notifications.

In general, I'm not usually in favor of the over-use of async style code (e.g. Javascript), but sometimes it make things simpler, which is one of the central tenants of Python.

The IDF only supports increments of 1 second for scanning.

Finally, what do you make of @pfalcon's [assertion](https://github.com/micropython/micropython-esp32/issues/68#issuecomment-294939523) that BLE doesn't belong in networking?

> * In general the API needs further careful consideration. In particular it's quite different from the alternative implementation in the ubluepy module in the nRF5x code: micropython/micropython#3137 . Both that and this implementation of BLE have their strengths/weaknesses and ultimately we want to have an API that works cleanly on esp32, nRF5x and other controllers. It will take some time and effort to get there but what's done here is a good starting point.

I feel a bit silly that I didn't even think to model this code off of bluepy; rather obvious.

However, some stuff doesn't make sense for µPy.  For example, why have a `Scanner` class? For bluepy this make sense, since the constructor takes an argument to select which BT device to use, but in µPy it's very unlikely there is more than 1.  `Scanner` is also a bit memory heavyweight, since it stores all the results internally (and the [nRF5x](https://github.com/tralamazza/micropython/blob/master/nrf/modules/ubluepy/ubluepy_scanner.c#L60) code does this as well) -- my idea for scanning is to pass the data back to user code, and let the user decide what's worth keeping.  

----

Some of the above are fairly low-effort (e.g. attributes vs functions) others are quite a bit of effort.  Are all of these blockers?

MrSurly committed:
esp32/bluetooth_docs.md: Fixed formatting, added .disconnect

MrSurly committed:
esp32/network_bluetooth.c: Added disconnect() to GATTCConn

dpgeorge commented:
> Not sure blocking would be better, especially WRT scanning, since many use cases are looking for a particular device, and scanning can take a bit of time in noisy environments -- scanning can be canceled when the device you're looking for is found.

Yes I agree that scanning can take some time and there can be many (hundreds) of results.  But a lot of the time results are duplicated (ie the same device, same MAC, it'll just have a different RSSI value) and so a scan could automatically discard duplicates (or keep only the last one).

> Would it be preferable to use threads in a blocking-scan sort of implementation? Doesn't solve what's described above, but would allow the device to do other things while waiting for device scan.

No, let's not design an API that requires threading :)

> Some features have to be async (like the HR demo), since it relies on indications/notifications.

Not necessarily.  One could set up the services and characteristics, then sit in an infinite loop checking for any event and responding to them.  Similar to a traditional socket, a GATTS object could have a read() method which blocks until an event comes in, and a write() method to return data.  It could even have an accept() method which waits for an incoming connection, and ability to poll the connection for an event.  There are lots of parellels that can be drawn.  And the GATTS connection could be put in non-blocking mode, and raise OSError(EAGAIN), if you want to use that form of programming.  It could then fit into the general (u)asyncio framework, alongside a traditional socket (eg wifi) server.

> Finally, what do you make of @pfalcon's assertion that BLE doesn't belong in networking?

As said above, there are lots of parellels that can be drawn between BLE and sockets (eg ability to scan(), lots of parameters to config(), that GATTC does connect() to connect to a remote server).  And the "network" module has very little in it, usually just the interface classes (eg WLAN).  A bluetooth module would probably just have the Bluetooth interface class as well, so putting it in the "network" module is a fair thing to do.

> However, some stuff doesn't make sense for µPy. For example, why have a Scanner class? For bluepy this make sense, since the constructor takes an argument to select which BT device to use, but in µPy it's very unlikely there is more than 1.

There could be more than one BT controller on the MCU, but it's unlikely.  Still there is a standard pattern already to support this: with the WLAN class you can make a STA and an AP instance, and they are independent interfaces.  So my idea would be that you create a BT instance for the central role (GATTC) and one for the peripheral role (GATTS), eg:
```python
gattc = network.Bluetooth(network.BLE_CENTRAL)
gatts = network.Bluetooth(network.BLE_PERIPHERAL)

gattc.scan()
gatts.advertise(prod='prod', manuf='manuf')
```

> Scanner is also a bit memory heavyweight, since it stores all the results internally (and the nRF5x code does this as well) -- my idea for scanning is to pass the data back to user code, and let the user decide what's worth keeping.

Yeah, the scanner does block but can easily run out of memory if run for a long time.

> Some of the above are fairly low-effort (e.g. attributes vs functions) others are quite a bit of effort. Are all of these blockers?

It would be nice to merge this PR to have something to play with, to then have a clearer idea of what the API could look like (and getting the BLE stuff working on the esp32 is non-trivial, so it's great to have this work already done, thanks!).  But if it is merged in the current state then it's really subject to complete change of the API.

dpgeorge commented:
An example of how the heart-rate example could be rewritten in a socket-like way:
```python
gatts = network.Bluetooth(network.BLE_PERIPHERAL)
s1 = gatts.Service(0xaabb)
s2 = gatts.Service(0xaabb)
s1.add_char(0xccdd)
s2.add_char(0xccdd)
s1.listen()
s2.listen()
gatts.advertise(prod='mangoprod', manuf='mangocorp')
while True:
    for s in select.select([s1, s2], [], [])[0]:
        event = s.read_event()
        if event == b.READ:
            print('Read')
            s.write('ABCDEFG')
        elif event == b.WRITE:
            print ('Write', s.read())
        elif event == b.NOTIFY:
            value = s.read()
            print ('Notify', value)
```
Note that s1.listen() replaces s1.start(), and the select combined with s.read_event() replaces the callbacks.

tannewt commented:
Awesome work @MrSurly and @dpgeorge!

I just wanted to chime in on the `network` module question. I'd push to have it as a separate module because some ports will have bluetooth support but not wifi support (nrf52 for example). I prefer that a module exist or not rather than having classes within a module exist or not. That way you can make support charts like [this](https://circuitpython.readthedocs.io/en/stable/shared-bindings/index.html#support-matrix).

Thanks again, keep up the good work!

chrismas9 commented:
>  I prefer that a module exist or not rather than having classes within a module exist or not

I see the merit of your chart, but where do you draw the line? Should `Ethernet` and `Wi-Fi` be separate modules? What about `DAC` which is available on pyboard, but not pyblite? You would have to break `machine` into may modules due to the variation in peripheral availability between MCU's. I also prefer to have bitbang and hardware IO (eg I2C, SPI) in the same module like `machine`. Why put the same functionality into different modules because the underlying implementation is different?

Instead of yes or no your availability chart could list the module name if available or blank if unavailable.

MrSurly commented:
@dpgeorge @tannewt 

I've updated the first comment for this PR with a checklist, which presently is just a mashup of suggestions here.  I'm not heavily invested in the present API; I like it, but the API should match what exists in other µPy modules.  I would, however, only like to change it once, so let's hammer this out.

I'll take on the work of making the API whatever is decided upon, but the API shouldn't make some use-cases too onerous to implement.

I do feel that removing callbacks makes the device less-friendly for multitasking.  My own use case involves the "main" task doing sensor reads or other non BLE stuff, and relies heavily on the BT callbacks.

What about using `queue` objects instead of callbacks?

tannewt commented:
@chrismas9 I draw the line where I think functionality is commonly grouped amongst microcontrollers. The intent is to make it easier to understand the functionality available on a particular port. I would split Ethernet and Wifi myself since boards may offer one or the other. However, if there are higher level TCP classes for example, I'd separate out the shared functionality too.

I think its fine to have many, smaller modules. It makes it clearer what functionality code uses and reduces memory usage when the modules actually need to be loaded in. (`machine` not included because its built in.)

`bitbangio` is separate from `busio` to make it clear what implementation is used for ports where both are available. Its a different way to handle the same thing that MicroPython's peripheral index does.

MrSurly commented:
> I think its fine to have many, smaller modules. It makes it clearer what functionality code uses and reduces memory usage when the modules actually need to be loaded in. (machine not included because its built in.)

Right now, WiFi is started when you `import network,` since that's what it does at startup. Since `Bluetooth` is in `network`, it can be a possibly unwanted side-effect, even if it's just increased power consumption.

MrSurly committed:
esp32/network_bluetooth.c: All properties now functions
esp32/bluetooth_docs.md: All properties now functions

Atzingen commented:
Is the bluetooth implemented in the oficial firmware (esp32-20170716-v1.9.1-219-g3580284e.bin) for me to try it ?

MrSurly commented:
No, just a PR right now, but you can check out the branch the PR is based
on.

On Jul 16, 2017 11:26, "Gustavo Voltani von Atzingen" <
notifications@github.com> wrote:

> Is the bluetooth implemented in the oficial firmware
> (esp32-20170716-v1.9.1-219-g3580284e.bin) for me to try it ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/micropython/micropython-esp32/pull/86#issuecomment-315627441>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAYHCIJmUMI6Xh89tbgg6LnZlSPmLf0Gks5sOlXqgaJpZM4NZmY1>
> .
>

MrSurly committed:
esp32/network_bluetooth.c: Fix for bad char write
https://github.com/MrSurly/micropython-esp32/issues/2

MrSurly committed:
esp32/network_bluetooth.c: Fix for disconnect/reconnect
https://github.com/MrSurly/micropython-esp32/issues/1

MrSurly committed:
esp32/bluetooth_docs.md: Updated doc for bluetooth.conns()

rdagger commented:
@MrSurly  any ETA when your bluetooth PR will be merged?

MrSurly commented:
@rdagger 

I'm not sure if the plan is to:

1) Make API changes, 2) merge

- or -

1) Merge 2) Make API changes.

I was waiting for comments from @dpgeorge  & others per the planned changes in the top comment.

MrSurly committed:
esp32/network_bluetooth.c: Fixed duplicate conn obj on device disconnect

MrSurly committed:
Merge branch 'esp32' into dev-bluetooth

MrSurly committed:
esp32/Makefile: Modified to be more inline with esp32 branch

MrSurly committed:
updated code example to match current API

ruudk commented:
Since this PR is a bit out of sync with this repository, it took me a long time to get it working. For anybody that wants to try it out, this is how I got it working on a fresh machine:

```
python --version
# Python 2.7.13

# xtensa
sudo easy_install pip
sudo pip install pyserial
mkdir -p ~/esp
cd ~/esp

# Make sure you grab the older Xtensa:
wget https://dl.espressif.com/dl/xtensa-esp32-elf-osx-1.22.0-61-gab8375a-5.2.0.tar.gz
tar -xzf xtensa-esp32-elf-osx-1.22.0-61-gab8375a-5.2.0.tar.gz

# ESP-IDF
cd ~/esp
git clone https://github.com/espressif/esp-idf.git
cd ~/esp/esp-idf
git checkout 4ec2abbf23084ac060679e4136fa222a2d0ab0e8
# SHA comes from from https://github.com/MrSurly/micropython-esp32/blob/dev-bluetooth/esp32/Makefile#L38
git submodule update --init --recursive

# Add to profile
export ESPIDF=/Users/Ruud/esp/esp-idf
export IDF_PATH=~/esp/esp-idf
export PATH=$PATH:$HOME/esp/xtensa-esp32-elf/bin

# ESP32 with Bluetooth
git clone https://github.com/MrSurly/micropython-esp32.git ~/micropython-esp32-bluetooth
cd ~/micropython-esp32-bluetooth
git checkout dev-bluetooth

git submodule init lib/berkeley-db-1.xx
git submodule update

make -C mpy-cross

cd esp32
make

export PORT=/dev/tty.SLAB_USBtoUART
make erase
make deploy
picocom -b 115200 /dev/tty.SLAB_USBtoUART
```

ruudk committed:
Fix docs > `adv_enable` needs `ble_` prefix

MrSurly committed:
Merge pull request #7 from ruudk/patch-1

Fix docs > `adv_enable` needs `ble_` prefix

Merged, thank you.

v923z commented:
@MrSurly Do you think you could upload the current version of the firmware to somewhere? I would be interested in testing the bluetooth implementation, but haven't the time to set up the development environment. Thanks.

MrSurly commented:
@v923z 

I put a release here, buuuut ... this is very alpha.

https://github.com/MrSurly/micropython-esp32/releases/tag/dev-bluetooth-A1

v923z commented:
@MrSurly 
Many thanks!

MrSurly committed:
beginning to update this branch. currently broken

diogoviannaaraujo commented:
Since this looks the most attractive high level BLE API I've found I think I'll be implementing a complex(for me) prototype using this, my first dig in micropython and been 7 years i've touched a python code, hope I can provide useful feedback.

I too think bluetooth deserves to live outside the network module, and use network maybe for tcp over bluetooth. Maybe someday I get confident enough to try some PRs

Beautiful work btw 👍

MrSurly commented:
The IDF implementation has changed considerably.  Closing this, will submit a new PR when new code is ready.
